torch/_inductor/fx_passes/reinplace.py

```diff
@@ -1,24 +1,331 @@
+    # realized as the output is a user of the input
+    if any(
+        user.target is aten.copy_.default and user.args[0] is inp for user in node.users
+    ):
+        return True
+
+    # Otherwise, assume fusions will make functional variants profitable
+    return False
+
+
+def decompose_generalized_scatter(graph: torch.fx.Graph) -> None:
+    """Replace _generalized_scatter with normal aten ops"""
+    for node in graph.nodes:
+        if node.target not in (_generalized_scatter, _inplace_generalized_scatter):
+            continue
+
+        use_mutation = (
+            node.target is _inplace_generalized_scatter
+            or scatter_always_uses_mutation(node)
+        )
+
+        with graph.inserting_before(node):
+            if use_mutation:
+                new_node = _decompose_scatter_mutating(graph, node)
+            else:
+                new_node = _decompose_scatter_functional(graph, node)
+
+        node.replace_all_uses_with(new_node)
+        graph.erase_node(node)
+
+
+def canonicalize_view_scatter_ops(graph: torch.fx.Graph) -> None:
+    """
+    This canonicalizes view scatter ops into a generalized form, defined as:
+      def scatter(inp, src, views):
+        tmp = inp.clone()
+        for view in views:
+          tmp = view(tmp)
+        tmp.copy_(src)
+
+    We also fuse consecutive view scatter ops of the form
+        a = scatter(view2(self), src, [view1])
+        b = scatter(self, a, [view2])
+    which can be rewritten as
+        b = scatter(self, src, [view2, view1])
+        a = view2(b)
+
+    This is both more efficient as we only do a single scatter, and also
+    easier to reinplace since there is only one use of `self`
+    """
+
+    node_to_view_base: Dict[torch.fx.Node, torch.fx.Node] = {}
+    node_to_view_op: Dict[torch.fx.Node, List[ViewOp]] = defaultdict(list)
+
+    def handle_views(node: torch.fx.Node):
+        inp = node.args[0]
+        node_to_view_base[node] = node_to_view_base.get(inp, inp)
+        node_to_view_op[node] = [
+            *node_to_view_op[inp],
+            ViewOp(
+                node.target,
+                args=node.args[1:],
+                kwargs=node.kwargs,
+            ),
+        ]
+
+    def handle_view_scatter(node: torch.fx.Node):
+        assert len(node.args) >= 2
+        inp, src = node.args[:2]
+
+        scatter_view_op = ViewOp(
+            _SCATTER_OP_TO_VIEW[node.target],
+            args=node.args[2:],
+            kwargs=node.kwargs,
+        )
+
+        if can_fuse := src.target is _generalized_scatter:
+            src_inp, src_src, src_scatter_view_op = src.args
+
+            inp_base = node_to_view_base.get(inp, inp)
+            src_base = node_to_view_base.get(src_inp, src_inp)
+            can_fuse = inp_base is src_base and node_to_view_op[src_inp] == [
+                *node_to_view_op[inp],
+                scatter_view_op,
+            ]
+
+        if not can_fuse:
```

Comment:
I don't quite understand this part. What happens here if you simply have a view -> view -> scatter? In this case we cannot fuse, but we are creating a generalized_scatter with just one view rather than two? Shouldn't we also add `*node_to_view_op[src]`?

Reply:
Lets say for example we have:
```python
def fn(a, src):
    a_view = a.as_strided(new_size, new_stride)
    a_view_updated = aten.slice_scatter(a_view, src, 1, 10, None)
    # ... use a_view_updated
```
IIUC your suggestion is to canonicalize this as:
```python
def fn(a, src):
    a_updated = _generalized_scatter(a, src, [
        ViewOp(aten.as_strided, (new_size, new_stride), {}), 
        ViewOp(aten.slice, (1, 10, None), {})
    ])
    a_view_updated = a_updated.as_strided(new_size, new_stride)
    # ...
```
where the scatter op is now in terms of the base tensor instead of the original scatter op argument. 

In this (cherry-picked) example it's rather problematic because `as_strided` forces the buffer to be realized and the scatter to be handled via mutation, whereas the original `slice_scatter` doesn't require mutation. Also even if the `as_strided` only views a subset of `a`, this version would still need to copy the entire buffer.

As written, this PR instead canonicalizes the example as:
```python
def fn(a, src):
    a_view = a.as_strided(new_size, new_stride)
    a_view_updated = _generalized_scatter(a_view, src, [ViewOp(aten.slice, (1, 10, None), {})])
    # ...
```

Which we can directly translate back to the original `slice_scatter` and so no buffer gets realized.